pageserver/src/tenant/layer_map/historic_layer_coverage.rs

```diff
@@ -0,0 +1,413 @@
+    map.insert(7..8, 120..300, "Layer 3".to_string(), true);
+
+    // Add wide and short layer
+    map.insert(0..9, 130..199, "Layer 4".to_string(), true);
+
+    // Add wide layer taller than some
+    map.insert(0..9, 140..201, "Layer 5".to_string(), true);
+
+    // Add wide layer taller than all
+    map.insert(0..9, 150..301, "Layer 6".to_string(), true);
+
+    // After layer 4 insertion
+    let version = map.get_version(135).unwrap();
+    assert_eq!(version.image_coverage.query(0), Some("Layer 4".to_string()));
+    assert_eq!(version.image_coverage.query(1), Some("Layer 1".to_string()));
+    assert_eq!(version.image_coverage.query(2), Some("Layer 4".to_string()));
+    assert_eq!(version.image_coverage.query(4), Some("Layer 2".to_string()));
+    assert_eq!(version.image_coverage.query(5), Some("Layer 4".to_string()));
+    assert_eq!(version.image_coverage.query(7), Some("Layer 3".to_string()));
+    assert_eq!(version.image_coverage.query(8), Some("Layer 4".to_string()));
+
+    // After layer 5 insertion
+    let version = map.get_version(145).unwrap();
+    assert_eq!(version.image_coverage.query(0), Some("Layer 5".to_string()));
+    assert_eq!(version.image_coverage.query(1), Some("Layer 5".to_string()));
+    assert_eq!(version.image_coverage.query(2), Some("Layer 5".to_string()));
+    assert_eq!(version.image_coverage.query(4), Some("Layer 5".to_string()));
+    assert_eq!(version.image_coverage.query(5), Some("Layer 5".to_string()));
+    assert_eq!(version.image_coverage.query(7), Some("Layer 3".to_string()));
+    assert_eq!(version.image_coverage.query(8), Some("Layer 5".to_string()));
+
+    // After layer 6 insertion
+    let version = map.get_version(155).unwrap();
+    assert_eq!(version.image_coverage.query(0), Some("Layer 6".to_string()));
+    assert_eq!(version.image_coverage.query(1), Some("Layer 6".to_string()));
+    assert_eq!(version.image_coverage.query(2), Some("Layer 6".to_string()));
+    assert_eq!(version.image_coverage.query(4), Some("Layer 6".to_string()));
+    assert_eq!(version.image_coverage.query(5), Some("Layer 6".to_string()));
+    assert_eq!(version.image_coverage.query(7), Some("Layer 6".to_string()));
+    assert_eq!(version.image_coverage.query(8), Some("Layer 6".to_string()));
+}
+
+/// Wrapper for HistoricLayerCoverage that allows us to hack around the lack
+/// of support for retroactive insertion by rebuilding the map since the
+/// change.
+///
+/// Why is this needed? We most often insert new layers with newer LSNs,
+/// but during compaction we create layers with non-latest LSN, and during
+/// GC we delete historic layers.
+///
+/// Even though rebuilding is an expensive (N log N) solution to the problem,
+/// it's not critical since we do something equally expensive just to decide
+/// whether or not to create new image layers.
+/// TODO It's not expensive but it's not great to hold a layer map write lock
+///      for that long.
```

Comment:
I should probably address this before merging

Reply:
One solution is to change `HistoricLayerCoverage::historic` from `BTreeMap` into a `rpds::RedBlackTreeMapSync`. Then during rebuild we can clone the entire layer map in O(1), update the clone (without holding lock on the original), and then swap it with the original.

Though I feel like there should be a simpler solution ..

Reply:
I see 3 options:
1. use rpds
2. keep it a BTreeMap, clone the whole thing on every rebuild
3. switch to sorted Vec, use Vec::append during rebuild, which uses memcpy

Tradeoffs (cons):
1. rpds: Queries are 70% slower
2. BTreeMap: **Every** rebuild clones the entire map at least
3. Vec: Write lock held for 1 vec memcpy (30 microseconds on my pc)

I'd go with option 2 for now because it's the smallest change from current code. We can follow up later and try other options if necessary.

@hlinnaka @knizhnik This would be a big change (I'll need to refactor the RwLocks), so let me know if you see easier solutions or if you'd prefer to review this separately.

Reply:
> Even though rebuilding is an expensive (N log N) solution to the problem,
it's not critical since we do something equally expensive just to decide
whether or not to create new image layers.

I do not completely agree with it. mage layers can be created relatively rare (especially with partial image layer approach). But rebuilding it after adding each delta layer seems to be expensive.

May be I missed something,  but as you mentioned - we usually adding layers with  newer LSNs,
why then you are rebuilding index after adding each delta layer?

Reply:
Tldr: We currently only do expensive rebuilds just to remove layers after GC. I think we can avoid that by just deleting everyting in `HistoricLayerCoverage::historic` that's before `latest_gc_cutoff_lsn`. If any layer is older than the cutoff and still needed for another branch, we can keep that layer in a separate layer map for that branch. I'll need to read some code to verify my logic.

Reply:
Created follow-up issue